File: backend/app/models/billing_report.py
# Copyright (c) 2026 Pablo Health, LLC. Licensed under AGPL-3.0.

"""API models for the practice's financial report.

Mirrors ``app.payments.reporting`` the way ``BalanceResponse`` mirrors
``app.payments.balance.BalanceSummary`` — the domain layer stays plain
dataclasses, and this is the typed shape the route hands back. Ids and
amounts only: no client, no payer contact detail, nothing clinical.
"""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class AgingBucketResponse(BaseModel):
    label: str
    count: int
    cents: int


class AgingReportResponse(BaseModel):
    buckets: list[AgingBucketResponse]


class PayerMixEntryResponse(BaseModel):
    payer_id: str
    payer_name: str
    billed_cents: int
    collected_cents: int


class PayerMixReportResponse(BaseModel):
    entries: list[PayerMixEntryResponse]


class CollectionsRateResponse(BaseModel):
    """The rate is ``collected / (billed - contractual_adjustment -
    write_off)``, left to the caller: cents divide exactly and percentages
    do not, so the API stays exact and rounds nowhere."""

    billed_cents: int
    collected_cents: int
    contractual_adjustment_cents: int
    write_off_cents: int


class PayerLagResponse(BaseModel):
    payer_id: str
    payer_name: str
    claim_count: int
    median_days: float
    p90_days: float


class ClaimPaymentLagReportResponse(BaseModel):
    by_payer: list[PayerLagResponse]


class BillingReportResponse(BaseModel):
    """The four sections together, for the window (or as-of date) given."""

    from_date: date
    to_date: date
    aging: AgingReportResponse
    payer_mix: PayerMixReportResponse
    collections_rate: CollectionsRateResponse
    claim_payment_lag: ClaimPaymentLagReportResponse

File: frontend/src/types/billingReport.ts
// Copyright (c) 2026 Pablo Health, LLC. Licensed under AGPL-3.0.

/**
 * The practice's financial report API types.
 *
 * Mirrors `app.models.billing_report` field for field. Every rate here is
 * left as cents on the wire — `collections_rate` carries the four figures a
 * rate divides out of, not a percentage — so rounding happens exactly once,
 * in the component that renders it.
 */

export interface AgingBucketResponse {
  label: string
  count: number
  cents: number
}

export interface AgingReportResponse {
  buckets: AgingBucketResponse[]
}

export interface PayerMixEntryResponse {
  payer_id: string
  payer_name: string
  billed_cents: number
  collected_cents: number
}

export interface PayerMixReportResponse {
  entries: PayerMixEntryResponse[]
}

/** `collected / (billed - contractual_adjustment - write_off)` is the rate. */
export interface CollectionsRateResponse {
  billed_cents: number
  collected_cents: number
  contractual_adjustment_cents: number
  write_off_cents: number
}

export interface PayerLagResponse {
  payer_id: string
  payer_name: string
  claim_count: number
  median_days: number
  p90_days: number
}

export interface ClaimPaymentLagReportResponse {
  by_payer: PayerLagResponse[]
}

export interface BillingReportResponse {
  from_date: string
  to_date: string
  aging: AgingReportResponse
  payer_mix: PayerMixReportResponse
  collections_rate: CollectionsRateResponse
  claim_payment_lag: ClaimPaymentLagReportResponse
}
